### PR TITLE
Add unit test for `get_matrix_a_star`

### LINKED ISSUE #78: Add unit test: matrix_a_star: `get_matrix_a_star`
The function `get_matrix_a_star` does not have a unit test. It should have one.

This will involve patching and mocking, as it's a parent function of several child functions. As examples of how to do this, see:

- `test_get_deltas` in `...tests / indicator_matrix / test_indicator_matrix.py`
- `test_get_match_scores` in `... / tests / match_scores / test_match_scores.py`

Essentially, you want to use the real input arguments for the parent function (which will be the same input arguments as for the first child function, so can be turned into a fixture and moved to a module-specific `conftest.py` script). Then, you patch all of the child functions' return values, so they will give the correct input to the next child function. This means you can then test the parent function by asserting that each child function was called once (only) with the correct input values.

### PR DESCRIPTION
**Pull request by mary-cleaton**
## Description

Add unit test for `get_matrix_a_star`. This involved moving a lot of stuff to fixtures stored in either the module or the whole-repo `conftest.py` script.

closes #78 

## Type of change

*You can delete options that are not relevant.*

- [ ] Bug fix - *non-breaking change*
- [x] New feature - *non-breaking change*
- [ ] Breaking change - *backwards incompatible change, changes expected behaviour*
- [x] Non-user facing change, structural change, dev functionality, docs ...

## Checklist:

- [x] I have performed a self-review of my own code.
- [x] I have commented my code appropriately, focusing on explaining my design decisions (explain why, not how).
- [x] I have made corresponding changes to the documentation (comments, docstring, etc.. )
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have updated the change log.

<br>

#  Peer review
Any new code includes all the following:

- **Documentation**: docstrings, comments have been added/ updated.
- **Style guidelines**: New code conforms to the project's contribution guidelines.
- **Functionality**: The code works as expected, handles expected edge cases and exceptions are handled appropriately.
- **Complexity**: The code is not overly complex, logic has been split into appropriately sized functions, etc..
- **Test coverage**: Unit tests cover essential functions for a reasonable range of inputs and conditions. Added and existing tests pass on my machine.

### Review comments
Suggestions should be tailored to the code that you are reviewing. Provide context.
Be critical and clear, but not mean. Ask questions and set actions.
<details><summary>These might include:</summary>

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented
  - Do the tests effectively assure that it
  works correctly? Are there additional edge cases/ negative tests to be considered?
- code style improvements (could the code be written more clearly?)
</details>
<br>

*Further reading: [code review best practices](https://best-practice-and-impact.github.io/qa-of-code-guidance/peer_review.html)*

### CODE REVIEW

File: tests/matrix_a_star/test_matrix_a_star.py [hunk 1]
```diff
@@ -89,9 +104,71 @@ def test_calculate_r():
...
+    mock_make_matrix_a,
+    mock_calculate_q,
+    mock_multiply_vectors_by_s,
+    mock_calculate_r,
+    mock_make_matrix_a_star,
+    spark_mock,
+    compare_deltas_output,
+    calculate_njklm_values_output,
+    make_matrix_a_output,
+):
```
Comment (mary-cleaton) [inline tests/matrix_a_star/test_matrix_a_star.py]: Here and throughout (including `conftest.py` scripts), add type hints.

File: tests/matrix_a_star/conftest.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
+import numpy as np
+import pandas as pd
+import pytest
+
+
+@pytest.fixture(scope="module")
+def calculate_njklm_values_output():
```
Comment (danasebo) [inline tests/matrix_a_star/conftest.py]: please add output type hint to this function

File: conftest.py [hunk 3]
```diff
@@ -36,3 +37,51 @@ def spark_mock():
     spark_mock.mode.return_value = spark_mock
     spark_mock.save.return_value = None
     return spark_mock
+
+
+@pytest.fixture(scope="module")
+def compare_deltas_output(spark):
```
Comment (danasebo) [inline conftest.py]: please add type hints to this function

File: tests/indicator_matrix/test_indicator_matrix.py [hunk 4]
```diff
@@ -112,7 +111,7 @@ def test_calculate_sorensen_dice(
     ch.assert_df_equality(df1=test_output, df2=expected_output, ignore_row_order=True)
 
 
-def test_compare_deltas(spark):
+def test_compare_deltas(spark, compare_deltas_output):
```
Comment (danasebo) [inline tests/indicator_matrix/test_indicator_matrix.py]: please add type hints to this function

File: tests/matrix_a_star/conftest.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+            "N_forename_3_forename_1",
+            "N_forename_3_forename_2",
+            "N_forename_3_forename_3",
+        ],
+    )
+
+
+@pytest.fixture(scope="module")
+def make_matrix_a_output():
+    """
```
Comment (danasebo) [inline tests/matrix_a_star/conftest.py:46]: please add output type hint to this function

File: tests/matrix_a_star/test_matrix_a_star.py [hunk 6]
```diff
@@ -50,9 +51,23 @@ def test_calculate_b():
     assert test_output == expected_output
 
 
-@pytest.mark.skip(reason="test shell")
-def test_calculate_njklm_values():
-    pass
+def test_calculate_njklm_values(
+    spark, compare_deltas_output, calculate_njklm_values_output
```
Comment (danasebo) [inline tests/matrix_a_star/test_matrix_a_star.py]: please add type hints to this function

**Review by danasebo — CHANGES_REQUESTED** [review]
looks good to me, but I want to re-review once this branch is rebased on PR#68 and PR#73 after they are merged into develop.

**Comment by mary-cleaton** [comment]
Draft until rebased.

**Comment by mary-cleaton** [comment]
Rebased to `develop`, since #73 has been merged. Also, corrected a few missing type hints and re-ran pre-commit hooks.

File: tests/matrix_a_star/test_matrix_a_star.py (hunk 1 above)
Comment (mary-cleaton) [reply tests/matrix_a_star/test_matrix_a_star.py]: Added type hints here.

File: tests/matrix_a_star/conftest.py (hunk 2 above)
Comment (mary-cleaton) [reply tests/matrix_a_star/conftest.py]: Done.

File: conftest.py (hunk 3 above)
Comment (mary-cleaton) [reply conftest.py]: Done.

File: tests/indicator_matrix/test_indicator_matrix.py (hunk 4 above)
Comment (mary-cleaton) [reply tests/indicator_matrix/test_indicator_matrix.py]: Done.

File: tests/matrix_a_star/conftest.py (hunk 5 above)
Comment (mary-cleaton) [reply tests/matrix_a_star/conftest.py:46]: Done.

File: tests/matrix_a_star/test_matrix_a_star.py (hunk 6 above)
Comment (mary-cleaton) [reply tests/matrix_a_star/test_matrix_a_star.py]: Done.

**Comment by mary-cleaton** [comment]
Resolved conflicts. Ready for final review by @danasebo.

File: tests/matrix_a_star/conftest.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+            "N_forename_3_forename_3",
+        ],
+    )
+
+
+@pytest.fixture(scope="module")
+def make_matrix_a_output() -> np.array:
+    """
+    Sets up the expected ouput for make_matrix_a, which is also the test input
+    for  and used by test_get_matrix_a_star.
```
Comment (danasebo) [inline tests/matrix_a_star/conftest.py]: little thing - two spaces instead of one after `for` (`grep` check confirms this is the only occurrence in the whole repo!)

Comment (mary-cleaton) [reply tests/matrix_a_star/conftest.py]: Ack! Fixed.

**Review by danasebo — APPROVED** [review]
looks good to me - everything works as expected and the test coverage is sufficient, many thanks!

File: tests/matrix_a_star/test_matrix_a_star.py (hunk 1 above)
Comment (danasebo) [reply tests/matrix_a_star/test_matrix_a_star.py]: I'm happy with the resolution, closing the thread now.
